numba/tests/test_types.py

```diff
@@ -235,6 +235,115 @@ def test_from_bidwidth(self):
+              [('flat', 0, 1, [0])])
+        self.assertFalse(ty.need_shaped_indexing)
+        ty = types.NumpyNdIterType((a, a))
+        check(ty, (f32, f32), 1, "C",
+              [('flat', 0, 1, [0, 1])])
+        self.assertFalse(ty.need_shaped_indexing)
+        ty = types.NumpyNdIterType((a, e, e, c64))
+        check(ty, (f32, i16, i16, c64), 1, "C",
+              [('flat', 0, 1, [0]), # a
+               ('0d', 0, 0, [1, 2]), # e, e
+               ('scalar', 0, 0, [3]), # c64
+               ])
+        self.assertFalse(ty.need_shaped_indexing)
+        ty = types.NumpyNdIterType((a, f))
+        check(ty, (f32, f32), 1, "C",
+              [('flat', 0, 1, [0]), ('indexed', 0, 1, [1])])
+        self.assertTrue(ty.need_shaped_indexing)
+        ty = types.NumpyNdIterType((f,))
+        check(ty, (f32,), 1, "C", [('indexed', 0, 1, [0])])
+        self.assertTrue(ty.need_shaped_indexing)
+
+        # 2-dim C-order iterator
+        ty = types.NumpyNdIterType((b,))
+        check(ty, (f32,), 2, "C", [('flat', 0, 2, [0])])
+        self.assertFalse(ty.need_shaped_indexing)
+        ty = types.NumpyNdIterType((b, c))
+        check(ty, (f32, c64), 2, "C", [('flat', 0, 2, [0]), ('indexed', 0, 2, [1])])
+        self.assertTrue(ty.need_shaped_indexing)
+        ty = types.NumpyNdIterType((d,))
+        check(ty, (i16,), 2, "C", [('indexed', 0, 2, [0])])
+        self.assertTrue(ty.need_shaped_indexing)
+        ty = types.NumpyNdIterType((b, c, d, d, e))
+        check(ty, (f32, c64, i16, i16, i16), 2, "C",
+              [('flat', 0, 2, [0]), # b
+               ('indexed', 0, 2, [1, 2, 3]), # c, d, d
+               ('0d', 0, 0, [4]), # e
+               ])
+        self.assertTrue(ty.need_shaped_indexing)
+        ty = types.NumpyNdIterType((a, b, c, d, d, f))
+        check(ty, (f32, f32, c64, i16, i16, f32), 2, "C",
+              [('flat', 1, 2, [0]), # a
+               ('flat', 0, 2, [1]), # b
+               ('indexed', 0, 2, [2, 3, 4]), # c, d, d
+               ('indexed', 1, 2, [5]), # f
+               ])
+        self.assertTrue(ty.need_shaped_indexing)
+
+        # 2-dim F-order iterator
+        ty = types.NumpyNdIterType((c,))
+        check(ty, (c64,), 2, "F", [('flat', 0, 2, [0])])
+        self.assertFalse(ty.need_shaped_indexing)
+        ty = types.NumpyNdIterType((c, b, c, f))
+        check(ty, (c64, f32, c64, f32), 2, "F",
+              [('flat', 0, 2, [0, 2]), # c, c
+               ('indexed', 0, 2, [1]), # b
+               ('indexed', 0, 1, [3]), # f
+               ])
+        self.assertTrue(ty.need_shaped_indexing)
+        ty = types.NumpyNdIterType((b, c, c, d, d, a, e))
+        check(ty, (f32, c64, c64, i16, i16, f32, i16), 2, "F",
+              [('indexed', 0, 2, [0, 3, 4]), # b, d, d
+               ('flat', 0, 2, [1, 2]), # c, c
+               ('flat', 0, 1, [5]), # a
+               ('0d', 0, 0, [6]), # e
+              ])
+        self.assertTrue(ty.need_shaped_indexing)
+
```

Comment:
It's not only the depth that matters but the combination of various depths, layouts, etc.
(as for `np.arange`, keep in mind this is testing the types domain ;-))